src/writer.rs

```diff
@@ -0,0 +1,735 @@
+// Copyright 2021 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+//! This module writes Flattened Devicetree blobs as defined here:
+//! <https://devicetree-specification.readthedocs.io/en/stable/flattened-format.html>
+
+use std::collections::BTreeMap;
+use std::convert::TryInto;
+use std::ffi::CString;
+use std::mem::size_of;
+
+use thiserror::Error as ThisError;
+
+use crate::{FDT_BEGIN_NODE, FDT_END, FDT_END_NODE, FDT_MAGIC, FDT_PROP};
+
+#[derive(ThisError, Debug)]
+pub enum Error {
+    #[error("Properties may not be added after a node has been ended")]
+    PropertyAfterEndNode,
+    #[error("Property value size must fit in 32 bits")]
+    PropertyValueTooLarge,
+    #[error("Total size must fit in 32 bits")]
+    TotalSizeTooLarge,
+    #[error("Strings cannot contain NUL")]
+    InvalidString,
+    #[error("Attempted to end a node that was not the most recent")]
+    OutOfOrderEndNode,
+    #[error("Attempted to call finish without ending all nodes")]
+    UnclosedNode,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+const FDT_HEADER_SIZE: usize = 40;
+const FDT_VERSION: u32 = 17;
+const FDT_LAST_COMP_VERSION: u32 = 16;
+
+/// Interface for writing a Flattened Devicetree (FDT) and emitting a Devicetree Blob (DTB).
+///
+/// # Example
+///
+/// ```rust
+/// use vm_fdt::FdtWriter;
+///
+/// # fn main() -> vm_fdt::FdtWriterResult<()> {
```

Comment:
The only reason I put it into a function was for `?` as you mentioned - I'm not sure using `unwrap` in an example is ideal, since users will likely copy and paste it, but I can switch to `unwrap` if that's the preferred style.

Reply:
I think that makes sense. So to get the best of both worlds, what do you say about the following:
- rename the function to something like `fdt_example`
- call the function and unwrap (this should be as a comment such that the users don't see it in the documentation; this way if we did something weird in the example, the doc tests will fail.

```rust
/// # fn fdt_example() -> vm_fdt::FdtWriterResult<()> {
/// let mut fdt = FdtWriter::new(&[]);
//// .....
/// let dtb = fdt.finish(0x1000)?;
/// # Ok(())
/// # }
/// # let _ = fdt_example().unwrap();
```

Reply:
Sorry, I totally missed this followup comment - I updated the docs to match.